registry-plugin/main.go

```diff
@@ -0,0 +1,246 @@
+
+		// If the URL path starts with "/v2/" (exactly /v2/ indicating the first handshake request to confirm registry V2 API),
+		// it means that the request is intended for the Instill Artifact registry. In this case, the traffic is hijacked
+		// and directly relayed to the registry. Otherwise, if the URL path does not match any of these patterns,
+		// the traffic is passed through to the next handler.
+		if !strings.HasPrefix(req.URL.Path, "/v2/") {
+			h.ServeHTTP(w, req)
+			return
+		}
+
+		// Authenticate the user via docker login
+		var username, password string
+		if req.URL.Path == "/v2/" {
+			username, password, ok = req.BasicAuth()
+			if !ok {
+				// Challenge the user for basic authentication
+				w.Header().Set("WWW-Authenticate", `Basic realm="Restricted"`)
+				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+				return
+			}
+
+			// Validate the api key and the namespace authorization
+			if !strings.HasPrefix(password, "instill_sk_") {
+				writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+				return
+			}
+
+			ctx := metadata.AppendToOutgoingContext(ctx, "Authorization", fmt.Sprintf("Bearer %s", password))
+			if resp, err := mgmtPublicClient.ValidateToken(
+				ctx,
+				&mgmtPB.ValidateTokenRequest{},
+			); err == nil {
+				userUID := resp.UserUid
+				// Check if the login username is the same with the user id retrieved from the token validation response
+				if resp, err := mgmtPrivateClient.LookUpUserAdmin(
+					ctx,
+					&mgmtPB.LookUpUserAdminRequest{
+						Permalink: "users/" + userUID,
+					},
+				); err == nil {
+					if resp.User.Id != username {
+						fmt.Println(1)
+						writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+						return
+					}
+				} else {
+					logger.Error(err.Error())
+					writeStatusInternalError(req, w)
+					return
+				}
+
+				// To this point, if the req.URL.Path is "/v2/", return 200 OK to the client for login success
+				writeStatusOK(req, w)
+				return
+			}
+
+			fmt.Println(2)
+			writeStatusUnauthorized(req, w, "Instill AI user authentication failed")
+			return
+		}
+
+		// Docker image tag format:
+		// [registry]/[namespace]/[repository path]:[image tag]
+		// The namespace is the user uid or the organization uid
+		var namespace string
+		matches := regexp.MustCompile(`/v2/([^/]+).*`).FindStringSubmatch(req.URL.Path)
+		if len(matches) >= 2 {
+			namespace = matches[1]
+		} else {
+			writeStatusUnauthorized(req, w,
+				"Namespace is not found in the image name. Docker registry hosted in Instill Artifact has a format"+
+					"[registry]/[namespace]/[repository path]:[image tag]."+
+					"A namespace can be a user or organization ID.")
+			return
+		}
+
+		// If the username and the namespace is not the same,
+		// check if the namespace is an organisation name where the user has the membership
+		if username != "" && namespace != username {
```

Comment:
This will never be executed because `username` is only set within the previous condition block.